### DISCUSSION TITLE
How to use raw eCAL::CPublisher() to send an array of uint8 (unsigned char)?

### DISCUSSION THREAD

**Discussion by daldr-ntml** [eclipse-ecal/ecal#1640] [Q&A]
Hi, I am new to eCAL. I want to send inter-process messages which are binary data stored in arrays of uint8 (unsigned char).  Publisher and Subscriber are always on the same host and I don't want to use protocol buffers to avoid unnecessary overhead. The target system is 64-bit Ubuntu, using gcc 11.

I guess I need to use the raw eCAL::CPublisher() class.  Is there an example that shows its use (send and receive) please?

**Comment by rex-schilasky** [comment]
Hi @DavidA2014, welcome as new eCAL user :-). First of all, from the eCAL user API logic it doesn't matter if your processes running on the same node or not, it's just about publishing on a topic and subscribing it where ever you prefer.

Yes, you can use the raw binary CPublisher::Send API. Here you can find a [binary publisher](https://github.com/eclipse-ecal/ecal/blob/3923a1ddae8f1bd6ebdc044cad19ba5d5805d8f7/samples/cpp/pubsub/binary/binary_snd/src/binary_snd.cpp) and here the matching [subscriber](https://github.com/eclipse-ecal/ecal/blob/3923a1ddae8f1bd6ebdc044cad19ba5d5805d8f7/samples/cpp/pubsub/binary/binary_rec/src/binary_rec.cpp).

**Reply by daldr-ntml** [reply]
Hi @rex-schilasky thanks very much for your reply. The binary publisher/subscriber code is helpful.

**Reply by daldr-ntml** [reply]
Hi @rex-schilasky, I have a follow-up question. For testing purposes, I want to run my eCAL publisher and subscriber in separate threads of a single process. In this case, should each thread call eCAL::Initialize and eCAL::Finalize, or should these be called only once by the process?

**Reply by rex-schilasky** [reply]
Hi @DavidA2014 just call it one time for the whole process. If you want to subscribe to your own (process internal) publisher you need to [enable pub/sub loopback](https://eclipse-ecal.github.io/ecal/_api/function_ecal__util_8h_1a8bcb30275052a25eb7baa8cddf238a8d.html#_CPPv4N4eCAL4Util14EnableLoopbackEb).

**Reply by daldr-ntml** [reply]
Hi @rex-schilasky, your answer was super-helpful. Reception is now working. Thank you!

**Reply by daldr-ntml** [reply]
Hi @rex-schilasky, sorry for lots of questions. Do you have a sample that would help with sending a 64-bit integer?

**Reply by rex-schilasky** [reply]
Hello @DavidA2014, this is what ChatGPT answered on your question:

```c++

#include <ecal/ecal.h>
#include <iostream>
#include <thread>
#include <chrono>

int main(int argc, char **argv)
{
  // Initialize eCAL
  eCAL::Initialize(argc, argv, "int64_publisher");

  // Create a Publisher
  eCAL::CPublisher pub("int64_topic");

  // Prepare the message
  int64_t value = 1234567890123456789LL;

  // Publish the message
  while (eCAL::Ok())
  {
    pub.Send(reinterpret_cast<const char*>(&value), sizeof(value));
    std::this_thread::sleep_for(std::chrono::milliseconds(100));
  }

  // Finalize eCAL
  eCAL::Finalize();

  return 0;
}

```

Keep in mind that sending raw binary values is problematic when dealing with different operating system, computing languages or compiler versions. That's what google protocol buffers were made for.

**Reply by daldr-ntml** [reply]
Hi @rex-schilasky, thanks for your answer.
